src/devicecontroller/DefaultDeviceController.ts

```diff
@@ -522,6 +550,22 @@ export default class DefaultDeviceController implements DeviceControllerBasedMed
     return '';
   }
 
+  private getActiveDeviceId(kind: string): string | null {
+    /* istanbul ignore else */
+    if (this.activeDevices[kind] && this.activeDevices[kind].constraints) {
+      const activeDeviceMediaTrackConstraints =
+        this.activeDevices[kind].constraints.audio || this.activeDevices[kind].constraints.video;
+      const activeDeviceConstrainDOMStringParameters = (activeDeviceMediaTrackConstraints as MediaTrackConstraints)
+        .deviceId;
+      const activeDeviceId = (activeDeviceConstrainDOMStringParameters as ConstrainDOMStringParameters)
```

Comment:
- Can we reuse `getDeviceIdStr()`?
- @RidipDe @ltrung  In `getDeviceIdStr()` too, it's not a good idea to extract the `exact` property only.

   Public APIs `chooseAudioInputDevice` and `chooseVideoInputDevice` take the `Device` type as an argument. The `Device` type can be `string | MediaTrackConstraints | MediaStream | null`. Customers can call it with a non-exact consraint, e.g.

  ```
  // This will fail
  chooseAudioInputDevice({
     audio: {
        deviceId: { ideal: 'my-device-id }
     }
  })
  ```

  You can skip it in this PR but please create an issue for tracking. This code will also be broken by our change soon. We'll remove the `exact` keyword conditionally for the Safari 12.0.0 and 12.1.0 support.

Reply:
Good catch. I will write up a GitHub issue and work on it after this PR since this is an existing code (getActiveDeviceId is in Ridip's previous commit) and we want to get this in our next NPM push.

Reply:
Since our SDK was using the 'exact' keyword convention we decided to extract only the 'exact' property

Reply:
Since exact and ideal have the same union type, we can add additional if condition to check the case where we have 'ideal' property instead of the 'exact'
    exact?: string | string[];
    ideal?: string | string[];

Reply:
`if ((deviceMediaTrackConstraints.deviceId as ConstrainDOMStringParameters).ideal) {
        /* istanbul ignore else */
        if (
          (deviceMediaTrackConstraints.deviceId as ConstrainDOMStringParameters).ideal as string
        ) {
          return (deviceMediaTrackConstraints.deviceId as ConstrainDOMStringParameters)
            .ideal as string;
        }
      }`